src/install/install.zig

```diff
@@ -10089,162 +10104,466 @@ pub const PackageManager = struct {
         }
     }
 
-    /// - Arg is name and possibly version (e.g. "is-even" or "is-even@1.0.0")
-    /// - Find package that satisfies name and version
-    /// - Copy contents of package into temp dir
-    /// - Give that to user
-    fn preparePatch(manager: *PackageManager) !void {
-        const @"pkg + maybe version to patch" = manager.options.positionals[1];
-        const name: []const u8, const version: ?[]const u8 = brk: {
-            if (std.mem.indexOfScalar(u8, @"pkg + maybe version to patch", '@')) |version_delimiter| {
-                break :brk .{
-                    @"pkg + maybe version to patch"[0..version_delimiter],
-                    @"pkg + maybe version to patch"[version_delimiter + 1 ..],
-                };
-            }
-            break :brk .{
-                @"pkg + maybe version to patch",
-                null,
+    fn nodeModulesFolderForDependencyID(iterator: *Lockfile.Tree.Iterator, dependency_id: DependencyID) !?Lockfile.Tree.NodeModulesFolder {
+        while (iterator.nextNodeModulesFolder(null)) |node_modules| {
+            _ = std.mem.indexOfScalar(DependencyID, node_modules.dependencies, dependency_id) orelse continue;
+            return node_modules;
+        }
+
+        return null;
+    }
+
+    fn pkgDepIdForNameAndVersion(
+        lockfile: *Lockfile,
+        pkg_maybe_version_to_patch: []const u8,
+        name: []const u8,
+        version: ?[]const u8,
+    ) struct { PackageID, DependencyID } {
+        const name_hash = String.Builder.stringHash(name);
+
+        const strbuf = lockfile.buffers.string_bytes.items;
+
+        const dependency_id: DependencyID, const pkg_id: PackageID = brk: {
+            var buf: [1024]u8 = undefined;
+            const dependencies = lockfile.buffers.dependencies.items;
+
+            var matches_found: u32 = 0;
+            var maybe_first_match: ?struct { DependencyID, PackageID } = null;
+            for (dependencies, 0..) |dep, dep_id| {
+                if (dep.name_hash != name_hash) continue;
+                matches_found += 1;
+                const pkg_id = lockfile.buffers.resolutions.items[dep_id];
+                const pkg = lockfile.packages.get(pkg_id);
```

Comment:
`pkg_id` can be `invalid_package_id` if it was not resolved, which would cause a segfault here.

This should check that `pkg_id` != `invalid_package_id` and if so continue